From 0d537a76b7c2c2d5e3612adc93ae50cf0fc1fd82 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:41 -0400
Subject: [virt] kvm/vmx: Fix DR6 update on #DB exception

Message-id: <1406040016-3289-127-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86205
O-Subject: [RHEL7 PATCH v2 126/221] KVM: VMX: Fix DR6 update on #DB exception
Bugzilla: 1079841 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

According to the SDM, only bits 0-3 of DR6 "may" be cleared by "certain"
debug exception. So do update them on #DB exception in KVM, but leave
the rest alone, only setting BD and BS in addition to already set bits
in DR6. This also aligns us with kvm_vcpu_check_singlestep.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 8246bf52c75aa9b9b336a84f31ed2248754d0f71)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 25ed5a2..525a743 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -4869,7 +4869,8 @@ static int handle_exception(struct kvm_vcpu *vcpu)
   dr6 = vmcs_readl(EXIT_QUALIFICATION);
   if (!(vcpu->guest_debug &
         (KVM_GUESTDBG_SINGLESTEP | KVM_GUESTDBG_USE_HW_BP))) {
-   vcpu->arch.dr6 = dr6 | DR6_FIXED_1;
+   vcpu->arch.dr6 &= ~15;
+   vcpu->arch.dr6 |= dr6;
    kvm_queue_exception(vcpu, DB_VECTOR);
    return 1;
   }
-- 
1.7.1